hashintel/hash
# Add tracing crate for structured logging

## 🌟 What is the purpose of this PR?

- This adds the [tracing.rs](https://docs.rs/tracing/latest/tracing/) crate which we will use for logging and tracing.
- Simple configuration of it is handled through the CLI using `clap`
- Logs are pretty-printed to stderr by default, _and_ appended in JSON format (structured logging) to a file

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202547633973664/f) _(internal)_
- Related Previous PRs for inspiration
  - #228 
  - #224 
  - #243 

## 🔍 What does this change?

- Upgrades `clap` version
- Adds the following dependencies:
  - `tracing`
  - `tracing-appender`
  - `tracing-error`
  - `tracing-subscriber`
- Adds the `logging` module with
  - An `init_logger` function which configures the tracing layers, subscribers, etc.
  - An `Args` struct used by `clap` for logging configuration

## 📜 Does this require a change to the docs?

- At the moment we're not maintaining docs for CLI usage, so no. Clap should generate the docs from the structs

## ⚠️ Known issues

- When using `tracing::error!("{err}:?")` we get a nice pretty-printed output in stderr (good), but we do not get a nice output in the JSON output file. Ideally we should find a way to serialize the errors within the JSON
  - [Asana task to track this](https://app.asana.com/0/1201095311341924/1202592524805847/f) _(internal)_

## 🐾 Next steps

- Use the library as we go

## 🛡 What tests cover this?

- None

## ❓ How to test this?

1. Run the project as normal
2. See that logs are pretty now
3. See that the `log` folder was created and logs are in a JSON file
4. Repeat 1-3 with various args from the `Args` struct in the logging module. e.g.
```
cargo run --package hash-graph --bin hash-graph -- --log-folder log2 --log-file-prefix potato --log-format json 
```

## 📹 Demo
<img width="1043" alt="image" src="https://user-images.githubusercontent.com/25749103/178519872-da4f3a3c-1440-4a24-843a-78eba478cbfc.png">
<img width="885" alt="image" src="https://user-images.githubusercontent.com/25749103/178519811-2bd416a7-3b89-45b1-b3b3-2b561a22f9fa.png">
